platforms/core-runtime/messaging/src/test/groovy/org/gradle/internal/event/ListenerBroadcastTest.groovy

```diff
@@ -373,6 +376,55 @@ class ListenerBroadcastTest extends Specification {
         broadcast.size() == 0
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/31537")
+    def 'listeners may be added or removed by multiple threads'() {
+        def threads = 10
+        def listenerCount = 200
+        List<TestListener> listeners = (1..listenerCount)
+            .collect {
+                // stubbing will do, and mocks are too expensive
+                Stub(TestListener)
+            }
+
+        when:
+        def barrier = new CyclicBarrier(threads)
+        async {
+            (0..<threads).each {threadIndex ->
+                def listenersToAdd = listeners.toList().tap(List::shuffle)
+                start {
+                    barrier.await()
+                    listenersToAdd.each { listener ->
+                        broadcast.removeAll([listener])
+                        Thread.sleep(1)
+                        broadcast.addAll([listener])
+                        Thread.sleep(1)
+                        broadcast.remove(listener)
+                        Thread.sleep(1)
+                        broadcast.add(listener)
+                        Thread.sleep(1)
+                    }
+               }
+            }
+        }
+
+        then:
+        broadcast.size() == listenerCount
+
+        when:
+        def found = 0
+        broadcast.visitListeners {
+            assert listeners.remove(it)
+            found++
+        }
+
+        then:
+        found == listenerCount
+        listeners.empty
+
+        where:
+        n << (1..4)
```

Comment:
🤔  Is the `n` used anywhere or does it play some other role?

Reply:
Just to produce multiple runs of the test. Not sure there is another way to do that?

Reply:
If we are trying to compensate for the lack of certainty (which this test cannot provide), it's better to make the test itself more "intensive" than to run it multiple times. It's similar to why we run each performance test once, but it internally does everything it can to provide certainty, e.g. runs multiple batches of work.

I don't feel strongly about this, but it does feel weird to have an unused variable that forces the test to run multiple times